fe/fe-core/src/main/java/org/apache/doris/clone/Rebalancer.java

```diff
@@ -0,0 +1,96 @@
+ * 2. createBalanceTask: given a tablet, try to create a clone task for this tablet.
+ * 3. getToDeleteReplicaId: if the rebalance strategy wants to delete the specified replica,
+ * override this func to let TabletScheduler know in handling redundant replica.
+ * NOTICE:
+ * 1. Adding the selected tablets by TabletScheduler may not succeed at all. And the move may be failed in some other places.
+ * So the thing you need to know is, Rebalancer cannot know when the move is failed.
+ * 2. If you want to make sure the move is succeed, you can assume that it's succeed when getToDeleteReplicaId called.
+ */
+public abstract class Rebalancer {
+    // When Rebalancer init, the statisticMap is usually empty. So it's no need to be an arg.
+    // Only use updateLoadStatistic() to load stats.
+    protected Map<String, ClusterLoadStatistic> statisticMap = new HashMap<>();
+    protected TabletInvertedIndex invertedIndex;
+    protected SystemInfoService infoService;
+
+    public Rebalancer(SystemInfoService infoService, TabletInvertedIndex invertedIndex) {
+        this.infoService = infoService;
+        this.invertedIndex = invertedIndex;
+    }
+
+    public List<TabletSchedCtx> selectAlternativeTablets() {
+        List<TabletSchedCtx> alternativeTablets = Lists.newArrayList();
+        for (Map.Entry<String, ClusterLoadStatistic> entry : statisticMap.entrySet()) {
+            for (TStorageMedium medium : TStorageMedium.values()) {
+                alternativeTablets.addAll(selectAlternativeTabletsForCluster(entry.getKey(),
+                        entry.getValue(), medium));
+            }
+        }
+        return alternativeTablets;
+    }
+
+    // The return TabletSchedCtx should have the tablet id at least. {srcReplica, destBe} can be complete here or
+    // later(when createBalanceTask called).
+    protected abstract List<TabletSchedCtx> selectAlternativeTabletsForCluster(
+            String clusterName, ClusterLoadStatistic clusterStat, TStorageMedium medium);
+
+
+    public void createBalanceTask(TabletSchedCtx tabletCtx, Map<Long, PathSlot> backendsWorkingSlots,
+                                  AgentBatchTask batchTask) throws SchedException {
+        completePlan(tabletCtx, backendsWorkingSlots);
+        batchTask.addTask(tabletCtx.createCloneReplicaAndTask());
+    }
+
+    // Before createCloneReplicaAndTask, we need to complete the TabletSchedCtx.
+    // 1. If you generate {tabletId, srcReplica, destBe} in selectAlternativeTablets(), it may be invalid at
+    // this point(it may have a long interval between selectAlternativeTablets & createBalanceTask).
+    // You should check the moves' validation.
+    // 2. If you want to generate {srcReplica, destBe} here, just do it.
+    // 3. You should check the path slots of src & dest.
+    protected abstract void completePlan(TabletSchedCtx tabletCtx, Map<Long, PathSlot> backendsWorkingSlots)
```

Comment:
In the whole FE balance module, we don't use `Plan`. So I think we could rename it to `completeSchedCtx`?

Reply:
👌